### PR TITLE
[WIP] Implementation of the LCNN model

### PR DESCRIPTION
**Pull request by VIGNESHinZONE**
# Implementation of the LCNN Model

This the second PR on the implementation of the LCNN Model (in Pytorch and DGL). 
The featuriser for this model is implemented https://github.com/deepchem/deepchem/pull/2287 and I will add the doctest after this gets merged.

## Rsults
This model on average gets 0.43 RMSE error on the Platinum Adsorption Dataset(Consisting of 650 crystal adsorbate configuration). This model beats the accuracy proposed by the authors in the LCNN paper where they achieve 4.3 RMSE error given [here](https://pubs.acs.org/doi/10.1021/acs.jpcc.9b03370).

## Checklist

- [x] My code follows [the style guidelines of this project]
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by VIGNESHinZONE** [comment]
This is just a resubmission of this https://github.com/deepchem/deepchem/pull/2383

File: deepchem/models/torch_models/lcnn.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+      node_feats = conv(G, node_feats)
+    node_feats = self.Atom_wise_Conv(node_feats)
+    node_feats = self.Atom_wise_Lin(node_feats).sum(axis=1)
+    G.ndata['new'] = node_feats
+
+    y = dgl.mean_nodes(G, 'new')
+    return y
+
+
+class LCNNModel(TorchModel):
```
Comment (rbharath) [inline deepchem/models/torch_models/lcnn.py:350]: Can you add unit tests and ML correctness tests for this class? See

https://deepchem.readthedocs.io/en/latest/development_guide/coding.html#unit-tests
https://deepchem.readthedocs.io/en/latest/development_guide/coding.html#testing-machine-learning-models

for a guide on the tests that we'll need. In particular, we'll need an overfitting test and a reloading test for `LCNNModel`

File: deepchem/models/torch_models/lcnn.py [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+    -------
+    node_feats: torch.Tensor
+        Updated tensor.
+    """
+    mask = self.dropout(self.ones).view(layer.shape[0], 1).repeat(
+        1, layer.shape[1])
+    return mask * layer
+
+
+class LCNN(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/lcnn.py:253]: Can you add simple unit tests for this class?

File: deepchem/models/torch_models/lcnn.py [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+      raise ImportError("This class requires DGL to be installed.")
+    G.ndata['x'] = node_feats
+    G.update_all(fn.copy_u('x', 'm'), self.reduce_func)
+    X = self.conv_weights(G.ndata['X_site'])
+    X = torch.stack([self.batch_norm(X_i) for X_i in X])
+    node_feats = torch.stack([self.dropout(X_i).sum(axis=0) for X_i in X])
+    return node_feats
+
+
+class AtomWiseLinear(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/lcnn.py:92]: Can you add a unit test for this class?

File: deepchem/models/torch_models/lcnn.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
+import torch
+import torch.nn as nn
+from deepchem.models.torch_models.torch_model import TorchModel
+from deepchem.models.losses import L2Loss
+
+
+class LCNNBlock(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/lcnn.py:7]: Can you add a unit test?

**Review by rbharath — COMMENTED** [review]
A good start! The biggest thing we need is unit tests. In particular, we need save/reload tests and overfitting tests for `LCNNModel`. I've linked some reference below. Take a look at how save/reload tests and overfit tests are done for the other models to see examples of how this is done

**Comment by VIGNESHinZONE** [comment]
@rbharath I have added the changes along with the test functions(overfitting and reload test). For some reason, the checks are breaking.

**Review by rbharath — APPROVED** [review]
Ok looks good! I think we can go ahead and merge this in if you're ready.

**Comment by VIGNESHinZONE** [comment]
I have no further commits to be added to the model. I think this can be merged :)
